### ISSUE TITLE
TFGraph.Dropout?

### ISSUE THREAD

**Issue by cesarsouza** [migueldeicaza/TensorFlowSharp#82]
Hi there,

I've noticed that the ```TFGraph``` class includes many operations from the tf.nn python modules, such as ELU, [Relu](https://migueldeicaza.github.io/TensorFlowSharp/api/TensorFlow/TensorFlow.TFGraph.html#TensorFlow_TFGraph_Relu_TensorFlow_TFOutput_System_String_), and TanH.

However, I couldn't help but notice that the [tf.nn.dropout](https://www.tensorflow.org/api_docs/python/tf/nn/dropout) operation is not present there (yet?).

Being that this is quite a ubiquitous operation in most modern deep learning applications, I guess it is available somewhere else, but I couldn't find yet where its precise location is right now. If that is the case, sorry for the stupid question, but may I ask where it could be found?

Regards,
Cesar

**Comment by cesarsouza** [comment]
If dropout is really not available at this time, may I ask what would be the procedure for adding it in TensorFlowSharp?

I am guessing that one of the first steps would be to import the random operation definitions from 
[random_ops.py](https://github.com/tensorflow/tensorflow/blob/r1.2/tensorflow/python/ops/random_ops.py) if they aren't available yet either (or possibly from [random_ops.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/random_ops.cc) since that python script uses that seemed to be auto-generated from this file).

In order to do that, would it be necessary to make changes to OpDefs / OpGenerator, or would it be a matter of implementing those operations directly into OperationExtras.cs?

May I ask what would be the expected way to start working on this issue?

Regards,
Cesar

**Comment by cesarsouza** [comment]
Oh, actually I have just found out that [RandomNormal is defined inside the Graph class](https://github.com/migueldeicaza/TensorFlowSharp/blob/master/TensorFlowSharp/OperationsExtras.cs#L191). What is the rule for deciding whether to put an operation inside Graph or inside OperationsExtras?

**Comment by cesarsouza** [comment]
I guess I am still getting used to the code base. The Random Uniform is already being auto-generated by OpGenerator.

**Comment by cesarsouza** [comment]
Closing issue, as the code has been merged a while ago.

**Comment by basejn** [comment]
This implementation of dropout , prevents you from applying dropout to a batch with first dimention -1 (dynamic batch size) because you convert the dynamic tensor Shape  of the input tensor to a constant ([-1,w,h...]) 

```
public TFOutput Dropout (TFOutput x, TFOutput keep_prob, TFShape noise_shape = null, int? seed = null, string operName = null)
		{
			var scopeName = MakeName ("dropout", operName);

			using (var newScope = WithScope (scopeName)) {
				if (noise_shape == null)
					noise_shape = new TFShape (GetShape (x));                                        

				TFOutput shapeTensor = ShapeTensorOutput (noise_shape);

				// uniform [keep_prob, 1.0 + keep_prob)
				TFOutput random_tensor = keep_prob;
				random_tensor = Add (random_tensor, RandomUniform (shapeTensor, seed: seed, dtype: x.OutputType));

				// 0. if [keep_prob, 1.0) and 1. if [1.0, 1.0 + keep_prob)
				TFOutput binary_tensor = Floor (random_tensor);
				TFOutput ret = Mul (Div (x, keep_prob), binary_tensor);
				SetTensorShape (ret, GetShape (x));
				return ret;
			}
		}
```
A more general approach would be : 
`public TFOutput Dropout (TFOutput x, TFOutput keep_prob, int? seed = null, string operName = null)
		{
			var scopeName = MakeName ("dropout", operName);

			using (var newScope = WithScope (scopeName)) {				                               

				TFOutput shapeTensor = getShape(x)

				// uniform [keep_prob, 1.0 + keep_prob)
				TFOutput random_tensor = keep_prob;
				random_tensor = Add (random_tensor, RandomUniform (shapeTensor, seed: seed, dtype: x.OutputType));

				// 0. if [keep_prob, 1.0) and 1. if [1.0, 1.0 + keep_prob)
				TFOutput binary_tensor = Floor (random_tensor);
				TFOutput ret = Mul (Div (x, keep_prob), binary_tensor);
				SetTensorShape (ret, GetShape (x));
				return ret;
			}
		}`
